module/VuFind/src/VuFind/Connection/OauthServiceTrait.php

```diff
@@ -0,0 +1,132 @@
+<?php
+
+/**
+ * Helper trait for OAuth 2.0 connections.
+ *
+ * PHP version 8
+ *
+ * Copyright (C) Villanova University 2023.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * @category VuFind
+ * @package  Connection
+ * @author   Demian Katz <demian.katz@villanova.edu>
+ * @author   Brent Palmer <brent-palmer@icpl.org>
+ * @author   Maccabee Levine <msl321@lehigh.edu>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org
+ */
+
+namespace VuFind\Connection;
+
+/**
+ * Helper trait for OAuth 2.0 connections.
+ *
+ * Closely adapted from VuFind\DigitalContent\OverdriveConnector.
+ *
+ * @category VuFind
+ * @package  Connection
+ * @author   Demian Katz <demian.katz@villanova.edu>
+ * @author   Brent Palmer <brent-palmer@icpl.org>
+ * @author   Maccabee Levine <msl321@lehigh.edu>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org
+ */
+trait OauthServiceTrait
+{
+    protected $tokenData = null;
+
+    /**
+     * Authentiate via the OAuth Client Credentials grant type.
+     *
+     * @param string $oauthUrl           URL of thee OAuth service
+     * @param string $clientId           client_id for a client_credentials grant
+     * @param string $clientSecret       client_secret for a client_credentials grant
+     * @param bool   $forceNewConnection Force a new connection (get a new token)
+     *
+     * @return string token for the session or false
+     *     if the token request failed
+     *
+     * @link https://www.oauth.com/oauth2-servers/access-tokens/client-credentials/
+     */
+    public function authenticateWithClientCredentials(
+        $oauthUrl,
+        $clientId,
+        $clientSecret,
+        $forceNewConnection = false
+    ) {
+        $this->debug("connecting to API");
```

Comment:
I am new to traits and didn't realize some of the limitations when I started; it would be cleaner to just have this trait use, the LoggerAwareTrait, but traits don't do that.  It seems ugly for a trait to force another dependency on classes that include it?  Maybe I should just remove the debug() calls (copied from Overdrive) and change the errors to throw an exception?

Reply:
Ok I took that approach, although I can't stomach passing 'debug' as a string each time so I added helper methods.  That's even more boilerplate that makes me not love traits, but it works.